Adding unittests to nuget.client locals

mishra14 opened the pull request:
Adding unit tests to nuget.client/nuget.exe to check if the arguments passed to locals are translated correctly when passed to nuget.xplat/nuget.cire localsCommandRunner.

Fixes : https://github.com/NuGet/Home/issues/3379

//cc: @joelverhagen @emgarten @alpaix @rohit21agrawal @jainaashish @drewgil

alpaix committed:
Build NuGet client solution with VS "15" willow

This change resolves number of issues blocking VS15 build on a box with VS
"15" willow installed at non-default location:
- Discovery of msbuild 15.1
- Microsoft.Build.* dependency version
- VS15/VS14 facade package restore conflict
- Assembly references pointing to a default VisualStudio PublicAssemblies location
- Binding VisualStudioVersion to MSBuildToolsetVersion

Key changes:
- Added 'Debug VS14'/'Release VS14' **solution** configurations to control
  projects built for specific "toolset", i.e target VS version
- Added `HintPath` to Microsoft.Build.* assemblies pointing to current
  `MSBuildToolsPath`
- Added `VisualStudioInstallDir` property initialized with VS install path
  from registry key
- Moved common project properties to `common.props`
- Project files cleanup

Commit:
changing localsCOmmandRunner to add localsArgs and make code more testable

Commit:
changing localsCOmmandRunner to add localsArgs and make code more testable

Commit:
Added semi-unit test cases for locals args parsing in nuget.client

Commit:
Added semi-unit test cases for locals args parsing in nuget.client

Commit:
Fixing second unittest to generate setting from config file

Commit:
Fixing second unittest to generate setting from config file

joelverhagen commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetLocalsCommandTest.cs:
```diff
@@ -90,5 +94,88 @@ public void LocalsCommand_Success_ValidLocalResource_ListMessage(string args)
+
+                // Assert
+                mockLocalsCommandRunner.Verify(mock => mock.ExecuteCommand(It.Is<LocalsArgs>(l => l.List && !l.Clear &&
+                                                                                             l.Arguments.Count == 1 && l.Arguments[0] == arg &&
```
please limit lines to 120 to make it good looking in GitHub

joelverhagen commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs:
```diff
@@ -1017,5 +1017,19 @@ private static bool IsProjectJson(string configFileName)
             return string.Equals(Path.GetExtension(configFileName), ".json", StringComparison.OrdinalIgnoreCase);
         }
+
+        public static string CreateDummyConfigFile(string directoryPath)
```
Does this need to be in a shared location? Can't it just be in the unit test class above?

joelverhagen commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetLocalsCommandTest.cs:
```diff
@@ -90,5 +94,88 @@ public void LocalsCommand_Success_ValidLocalResource_ListMessage(string args)
+                // Assert
+                mockLocalsCommandRunner.Verify(mock => mock.ExecuteCommand(It.Is<LocalsArgs>(l => !l.List && l.Clear &&
+                                                                                             l.Arguments.Count == 1 && l.Arguments[0] == arg &&
+                                                                                             l.LogError == mockConsole.Object.LogError &&
```
`ReferenceEquals` instead of `==` shows the intent here a bit better, IMO. Also, I don't see you `Setup` the `LogError` and `LogInformation` projects, so they are just null, right?

joelverhagen commented:
Address comments then :shipit:

joelverhagen committed:
Make NoCache skip reading packages from the global packages folder (#856)

Fix https://github.com/NuGet/Home/issues/1406
Fix https://github.com/NuGet/Home/issues/3074

Commit:
Refactoring helper method from Util.cs to LocalsUtil.cs and refactoring code to look concise

Commit:
Refactoring helper method from Util.cs to LocalsUtil.cs and refactoring code to look concise

zhili1208 committed:
Fixed Invalid Project Url parse error

https://github.com/NuGet/Home/issues/3420

drewgillies committed:
Ensure version of NuGet.VisualStudio.Interop is 1.0.0.0 in order to not break VS templates.

joelverhagen committed:
Clear cache before running V3 mock server test (index.json is cached) (#867)

zhili1208 committed:
JsonData test fix

joelverhagen committed:
Simplify PackageRestoreManager, removing unused code (#865)

Use FallbackPackagePathResolver in more places
Move init.ps1 execution for UWP project.json to shared location
Include restored packages in restore result (for packages.config)
Move PowerShell command to execute a .ps1 to a shared location

Commit:
Merge branch 'dev-anmishr-locals-unittests' of https://github.com/NuGet/NuGet.Client into dev-anmishr-locals-unittests

mishra14 commented:
Merged: https://github.com/NuGet/NuGet.Client/commit/f67f6660547119689b559bf2b0725ac8a0176a60
